courses/advanced-ada/chapters/design_patterns.rst

```diff
@@ -960,3 +965,661 @@ This is the complete source-code:
+finalization will automatically release it, including both when an
+exception is raised and when the task is aborted.
+
+.. admonition:: In other languages
+
+    C++ programmers may be familiar with this technique under the name
+    "Resource Acquisition Is Initialization" (RAII).
+
+The idea is to define a limited controlled type that references a shared
+lock using a discriminant. Objects of the type are then declared within
+procedures and functions with a discriminant value designating the shared
+lock declared within the package. Such a type is called a *scope lock*
+because the elaboration of the enclosing declarative region |mdash| the
+scope |mdash| is sufficient to acquire the referenced lock.
+
+To define the :ada:`Scope_Lock` type, we declare it with a discriminant
+designating a :ada:`Mutual_Exclusion` object:
+
+.. code:: ada
+
+    with Semaphores;  use Semaphores;
+
+    with Ada.Finalization;
+
+    package Locks is
+
+       subtype Mutual_Exclusion is Binary_Semaphore
+         (Initially_Available => True);
+
+       type Scope_Lock (Lock : access Mutual_Exclusion) is
+         tagged limited private;
+
+    private
+
+       type Scope_Lock (Lock : access Mutual_Exclusion) is
+         new Ada.Finalization.Limited_Controlled with null record;
+
+       overriding procedure Initialize (This : in out Scope_Lock);
+       overriding procedure Finalize   (This : in out Scope_Lock);
+
+    end Locks;
+
+In the private part the type is fully declared as a controlled type
+derived from :ada:`Ada.Finalization.Limited_Controlled`, as shown below.
+We hide the fact that the type will be controlled because
+:ada:`Initialize` and :ada:`Finalize` are never intended to be called
+manually.
+
+Each overridden procedure simply references the semaphore object
+designated by the formal parameter's discriminant:
+
+.. code:: ada
+
+    with Ada.Finalization;
+
+    package body Locks is
+
+       procedure Initialize (This : in out Scope_Lock) is
+       begin
+          This.Lock.Seize;
+       end Initialize;
+
+       procedure Finalize (This : in out Scope_Lock) is
+       begin
+          This.Lock.Release;
+       end Finalize;
+
+    end Locks;
+
+The subprogram's sequence of statements will not execute until the lock is
+acquired, no matter how long that takes. When the procedure or function is
+done, for any reason, finalization will release the lock. The resulting
+user code is thus almost unchanged from the original sequential code:
+
+.. code:: ada
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Locks; use Locks;
+
+    package body P is
+
+       Mutex : aliased Mutual_Exclusion;
+
+       State : Integer := 0;
+
+       procedure Operation_1 is
+          S : Scope_Lock (Mutex'Access);
+       begin
+          State := State + 1;  -- for example...
+          Put_Line ("State is now" & State'Img);
+       end Operation_1;
+
+       procedure Operation_2 is
+          S : Scope_Lock (Mutex'Access);
```

Comment:
Just a question: How does S not getting treated as an unused variable? Is it the Finalization that "uses" the object?